### DISCUSSION TITLE
Filtering URL Parameter Formatting

### DISCUSSION THREAD

**Discussion by eito-fis** [mathesar-foundation/mathesar#232] [Maintainer discussions]
Summarizing some options we have for filter URL parameter formatting here.

## Default Form Format
The simplest solution available to us. We can use a django `Form` with two [ `MultipleChoiceField` ](https://docs.djangoproject.com/en/3.2/ref/forms/fields/#multiplechoicefield)s to validate and process the input. As the `MultipleChoiceField` takes its inputs as multiple parameters of the same name, the query string would look something like:

```
?filter_name=thing&filter_name=thing2&filter_name=thing3&filter_value=a&filter_value=b&filter_value=c
```
### Pros
* Simple to implement, mostly just works out of the box
* Simple to use, just `form.is_valid()` and `form.clean_data`
* Automatically handles and generates errors

### Cons
* Would require some modification to support different operators down the line
* Long URL
* Potential issues with types, unclear how to properly coerce from strings back into python types
* Filter names could collide with other keys (e.g. filtering by a field name called `sort` could collide with the query parameter for sorting)

## Django-Filter Format
As we use django-filter for filtering on our other endpoints (#219, #230), it would be nice to maintain consistency and use a similar format for this endpoint. Unfortunately, as django-filter always operates on top of `Queryset`s, we would have to manually replicate this behavior with a custom form field or validation function. This could look something like this (unzipped form):
```
?filter_names=thing1,thing2,thing3&filter_values=a,b,c
```
or, could look something like this (condensed form):
```
?filters=thing1:eq:a,thing2:eq:b,thing3:eq:c
```
with `:`s delimiting column name, operator and value. Given the support for operators and single field, I think the second form is cleaner. 

### Pros
* Consistent with other endpoints, and any others we make with django-filter
* Short URL, human readable
* Easy support for operators

### Cons
* Requires custom functions / handling
* Potential issues with types, unclear how to properly coerce from strings back into python types
* Potential issues with delimiter character - what do we do if a column name or value contains `:`?

## JSON Array Format
Differs from other endpoints, but is a nice combination of easy to implement and general enough to support our needs. Django forms already have a [`JSONField`](https://docs.djangoproject.com/en/3.2/ref/forms/fields/#jsonfield) that we could use, which covers most of the worries about types / escaping characters / validation. We would still have to parse the data afterwards, but it would be relatively easy. This format would look something like:
```
?filters=[['thing1', 'eq', 'a'], ['thing2', 'eq', 'b'], ['thing3', 'eq', 'c']] 
```
with arrays instead of objects for a compact representation.

### Pros
* Mostly automatic validation / cleaning
* Easy support for operators
* No worries about escaping characters or types

### Cons
* Inconsistent with other endpoint behavior
* Potentially less human readable URLs

The decision seems pretty clearly to be between the JSON format and the condensed django-filter format - which one we choose depends on how much we care about consistency. The fact that we can reply on existing tools when implementing the JSON format is really appealing to me, but I also would like the formatting to be consistent across endpoints. My preference would likely be the latter (so django-filter format), but this is maybe a broader question about how we want to style the API, and if we want to commit to the CSV style URL parameters. Could definitely see an argument for JSON if we just wanted to implement something fast for the MVP.

**Comment by mathemancer** [comment]
In the long run, I think we'll need to be able to handle negation, conjunction, and disjunction w.r.t. filters.  Note that once you have both conjunction and disjunction you also need something to handle association (grouping with parentheses).
1. @eito-fis have you considered if one of these will be best extendable in those directions?
2. @kgodey @pavish are you aware of any examples we could look at for this?

**Reply by mathemancer** [reply]
Edit:  It seems that the `.filter` method of models basically puts things together using Conjunctive Normal Form.  Is that the standard?  If so, my preference would be `column1=a,b&column1=c` for something like `(column1 = a OR column1 = b) AND column1 = c` (related to the UI discussion, though I don't think we should necessarily wed the webapp UI and the API "UI").  Not sure how negations would be handled for that.

**Reply by kgodey** [reply]
I don't think we can handle association with either syntax. To support more complex queries, we'll need some sort of query syntax that we can pass (see `ComplexFilterBackend` in this project's `README`: https://github.com/philipn/django-rest-framework-filters). Perhaps we should just implement that syntax to support more complex queries in the future.

In that case, we'd have something like `?filters="(title__icontains='Hobbit'&author__icontains='Tolkien')|title__icontains='Pooh'"` (which we'd then have to URL encode properly). I'm inclined to go with this option, since we'll need to implement a parser either way. It'll take longer to implement, but it'll be more extensible/flexible for the future.

**Reply by pavish** [reply]
I agree with @kgodey on this. We will need a query syntax and write our own parser (or find any good one).

For reference: [Microsoft's api guidelines for filters](https://github.com/microsoft/api-guidelines/blob/vNext/Guidelines.md#971-filter-operations)

**Reply by mathemancer** [reply]
We certainly can write any possible filter in Conjunctive Normal Form.  Basically, the `&` between query params makes the parentheses.  Translating into CNF would be a bit tricky if the UI doesn't guide that direction, so I wouldn't really recommend doing things that way unless the UI is in sync somewhat.

If we do want to support "textual" filters, I suggest making it something that will be pretty similar to the underlying SQL `WHERE` statement.

**Comment by kgodey** [comment]
@eito-fis I assume you have enough information here for an updated recommendation. Let us know if you have any more questions.

**Comment by eito-fis** [comment]
Thanks for all the input!

There seem to be two decisions to make.

### Normal Form
Whether we want to take in arbitrary conditions or conditions in conjunctive normal form. CNF is nice due to the reasons outline by @mathemancer above. Namely, the query params will stay simple since we only need one level of grouping, which could be achieved easily with `&`s:
```python
'?filters=name:eq:person&filter=age:gt:50,private:eq:true'  #(name == person) and (age > 50 or private)
```
However, CNF form comes with the downside of begin somewhat un-intuitive (or at least require more work to put together) than an arbitrary conditional. On the other hand, allowing for any format of conditional will probably require more effort to implement. If we did want to go this route, the [Microsoft recommendations](https://github.com/microsoft/api-guidelines/blob/vNext/Guidelines.md#971-filter-operations)(thanks @pavish) offer a good starting point we can extend with additional operators.
```python
'?filters=name eq "person" and (age gt 50 or private)'
```
My preference here is for the non-CNF format. It is going to be more work to implement, but I imagine filtering records is something that we'll do on a regular basis and seems like something worth investing time into.

### Operations
For now, the backend only supports filtering off equality but I expect we'd like to expand this functionality. Regardless of what form we decide to use, we'll have to choose what operators to support and how to map them back to SQL compatible statements. We could come up with some custom mappings here, but I agree with @kgodey in that the drf-filters styling seem like a good choice - django has already had to define [a mapping of operators to SQL](https://docs.djangoproject.com/en/3.2/ref/models/querysets/#field-lookups), so we might as well use them too.

---
All this to say, the drf-filters style params seem like the way to go. Pain points will be around parsing the actual query string and mapping the parsed equation to SQL, but it shouldn't be too bad.

**Reply by kgodey** [reply]
This recommendation makes sense to me.

We could also look into libraries like https://github.com/juliotrigo/sqlalchemy-filters to make the mapping to SQL easier.

**Reply by pavish** [reply]
Sounds good to me!

**Reply by mathemancer** [reply]
I also think this makes sense.  I agree with @kgodey that we should explore libraries for the parsing/mapping before implementing anything custom.  We're not the first to deal with this problem.

**Reply by eito-fis** [reply]
Just noticed `sqlalchemy-filters` has support for nested conditionals! Might still need to do some parsing first (maybe [pycond](https://github.com/axiros/pycond#parsing)?) but will definitely try it out.

**Comment by eito-fis** [comment]
Having had some more time to think about this and finally starting work on it, `sqlalchemy-filters` does seem like the ideal library for our use case. It has a relatively simple format (basically the JSON format that was suggested earlier), handles negation, conjunction, and disjunction and is explicitly built to work with sqlalchemy. However, it doesn't support sqlalchemy 1.4 :disappointed: 

Could I get some help on figuring out the way to move forward here? In my mind we can:
* Use [the fork that supports 1.4](https://github.com/juliotrigo/sqlalchemy-filters/pull/64) but seem unlikely to be merged upstream anytime soon
* Make our own fork that we patch for 1.4
* Rewrite our own filtering module in `db`, reusing `sqlalchemy-filters` general structure (but probably stripped down)
* Rewrite our own filtering module in `db` for the same format, but from scratch

Using the fork seems nicest just because it requires the least amount of additional work on our part, but I am also somewhat uncomfortable with being dependent on a fork that presumably isn't going to be closely maintained. Regardless, I don't have much experience with situations like this so advice would be appreciated.

**Reply by pavish** [reply]
My actions in such situations is often subjective. I would usually use the fork, since the fork has only few changes from the upstream, and the chances of it being merged is high.

However in this case, there hasn't been any work on the entire `sqlalchemy-filters` project since May 2020. And it seems to be maintained by a single person, which tends to the project being volatile.

I would raise an issue on the project, asking if it will be actively maintained, or if they would consider adding more maintainers so that we could contribute. We could also look for alternatives in the meanwhile.

If there is no response as such, it would be better if we wrote our own module.

@kgodey, @mathemancer What are your thoughts on this?

**Reply by kgodey** [reply]
I think we should fork `sqlalchemy-filters`, merge the patch that supports 1.4 in, and use our patched version. We can maintain our patched version as needed. I think that's better than writing our own module.

If the patch gets merged upstream, we can go back to using the actual `sqlalchemy-filters`.

**Reply by kgodey** [reply]
@mathemancer I'd like your sign off on my plan too, since you'll be helping maintain the fork.

**Reply by mathemancer** [reply]
I think it's a good plan.  I also like @pavish 's suggestion that we contact the maintainer to see if they're still working on the project.  It does seem to have been quite quiet for some time.

**Reply by eito-fis** [reply]
I've setup the fork [here](https://github.com/centerofci/sqlalchemy-filters/tree/models_to_tables). Unfortunately, its also setup to be used with the 1.x style queries, and not the 2.0 version queries we're currently using. It's actually quite easy to update if we pull out the auto-joining features that it has built it , but we should consider revisiting down the line if that's something we want to support.

**Reply by kgodey** [reply]
@eito-fis I don't follow – what's easy to update / what auto-joining features / what should we consider revisiting down the line?

**Reply by eito-fis** [reply]
@kgodey Sorry, will clarify:

It's easy to update most of the library to use the 2.0 style `selectable` queries we're using. But, the library also offers an auto-join feature where the user can supply different table names in the filter parameters to have them joined together before applying the filters. This feature uses specific portions of the old 1.x style queries that make it harder to port over so I stripped it out. Given that the current endpoint is for a single table I think this is fine, however this could be a nice feature to have down the line depending on our API's implementation of joins (or just to have in the library as an option).

**Reply by kgodey** [reply]
@eito-fis That makes sense, thank you. Could you create a ticket to track re-implementing joins so we don't forget about it? You can make it on the `sqlalchemy-filters` repo.

**Comment by mathemancer** [comment]
@kgodey What's our appetite for customizing our fork of `sqlalchemy-filters`?  Two more specific points:
- The library is set up such that you can't filter by a function of a column.  This is because it's basically designed for using with the SQLAlchemy ORM, and you'd never need to filter by a function of a column.  (by this, I mean something like filtering by `your_func(your_col) = 'a'`, rather than `your_col = 'a'`.  Extending the classes to support that won't work via inheritance; I'd need to get into the classes and extend their definitions in our `sqlalchemy-filters` source code to enable that.  It's needed for filtering by parts of emails and URIs (so far).
- On the ORM topic, the library has a bunch of complexity that's not needed for our purposes at all, since it's supporting the ORM query type.

My preference would be to rip that library apart a bit and cut down to just the parts we need, and customize them as needed.  This would let us more easily extend them in the future (because we wouldn't be stumbling over parts we don't need).  The downside would be that we would make our fork useless for folks who _do_ want the ORM parts of things, and we wouldn't be able to easily merge patches back upstream.  However, as you may recall, the upstream seems to be on a very slow development pace, so that may not be an issue.

**Reply by kgodey** [reply]
@mathemancer I'm fine with cutting down the library to the parts we need. I don't think we need to prioritize maintaining a usable fork. It's critical enough to Mathesar that I think customizing it to our own purposes to make it maintainable makes sense.
